File: kafka/golang/controllers/kafka_controller.go
package controllers

import (
	"net/http"
	"note-golang-kafka/services"

	"github.com/labstack/echo/v4"
)

type KafkaController interface {
	SendMessage(c echo.Context) error
}

type KafkaControllerImplementation struct {
	KafkaService services.KafkaService
}

func NewKafkaController(kafkaService services.KafkaService) KafkaController {
	return &KafkaControllerImplementation{
		KafkaService: kafkaService,
	}
}

func (controller *KafkaControllerImplementation) SendMessage(c echo.Context) error {
	textMessage := c.QueryParam("message")
	response := controller.KafkaService.SendMessage(c.Request().Context(), textMessage)
	return c.JSON(http.StatusOK, map[string]interface{}{
		"response": response,
	})
}

File: kafka/rust/src/controllers/kafka_controller.rs
use std::sync::Arc;

use axum::{extract::{Path, State}, http::StatusCode, response::IntoResponse};

use crate::{services::kafka_service::KafkaService, states::app_state::AppState};

pub async fn send_message(
    State(state): State<Arc<AppState>>,
    Path(message): Path<String>
) -> impl IntoResponse {
    let response = state.kafka_service.send_message(message).await;
    (StatusCode::OK, response)
}